KAFKA-12988 Asynchronous API support for RemoteLogMetadataManager add/update methods.

satishd opened the pull request:
- Added asynchronous API support for RemoeLogMetadataManager add/update/put methods.
 - Implemented the changes on default topic based RemoteLogMetadataManager.
 - Refactored the respective tests to cover the introduced asynchronous APIs. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-12988 Asynchronous API support for RemoteLogMetadataManager add/update methods.
 - Added asynchronous API support for RemoeLogMetadataManager add/update/put methods.
 - Refactored the respctive tests to cover enhanced APIs.

satishd commented:
@junrao This PR is rebased with trunk, please review and let me know your comments.

satishd commented:
@junrao gentle reminder to review the changes.

junrao commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java:
```diff
@@ -62,16 +63,17 @@
+     * @return a Future which will complete once this operation is finished.
      */
-    void addRemoteLogSegmentMetadata(RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException;
+    Future<Void> addRemoteLogSegmentMetadata(RemoteLogSegmentMetadata remoteLogSegmentMetadata) throws RemoteStorageException;
```
Should we use CompletableFuture instead of Future? With Future, there isn't much the caller could do.

satishd replied:
Updated the PR with the suggested change of having `CompletableFuture`.

junrao reviewed (commented):
@satishd : Thanks for the PR. Left a high level comment below.

satishd committed:
Added support for CompletableFuture instead of Future for RemoteLogMetadataManager.
 - Addressed review comments.
 - Added support for CompletableFuture instead of Future for RemoteLogMetadataManager.

satishd commented:
Thanks @junrao for your comment. Addressed the review comment with the latest commit.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/FuturePublishRecordMetadata.java:
```diff
@@ -0,0 +1,131 @@
+ *     <li>{@link TopicBasedRemoteLogMetadataManager#putRemotePartitionDeleteMetadata(RemotePartitionDeleteMetadata)}</li>
+ * </ul>
+ */
+class FuturePublishRecordMetadata implements Future<Void> {
```
This class seems never used?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -129,37 +132,44 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate segmen
-            // semantics.
-            consumerManager.waitTillConsumptionCatchesUp(recordMetadata);
+            CompletableFuture<RecordMetadata> produceFuture = new CompletableFuture<>();
+            producerManager.publishMessage(remoteLogMetadata, produceFuture);
```
It seems that it's more natural for producerManager to create a CompletableFuture and return it through publishMessage() rather than passing in a CompletableFuture from the caller.

satishd replied:
I was checking out a few other ways and passing future to the respective functions. But the latest code does not require that to be passed and it can be returned as you suggested.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -129,37 +132,44 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate segmen
+                    throw new KafkaException(e);
+                }
+                return null;
+            }).toCompletableFuture();
```
Is toCompletableFuture() needed since thenApplyAsync() returns CompletableFuture already?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -129,37 +132,44 @@ public void updateRemoteLogSegmentMetadata(RemoteLogSegmentMetadataUpdate segmen
+            producerManager.publishMessage(remoteLogMetadata, produceFuture);
+            return produceFuture.thenApplyAsync((Function<RecordMetadata, Void>) recordMetadata -> {
+                try {
+                    consumerManager.waitTillConsumptionCatchesUp(recordMetadata);
```
Hmm, this callback will be called from Producer's Sender thread and consumerManager.waitTillConsumptionCatchesUp() blocks until the timeout. This will block the Sender thread, which is not ideal.

satishd replied:
This is not invoked from Producer's sender thread as it is run as async. You can see the below stack trace. By default it uses common `ForkJoinPool` if it supports parallelism.

```
org.apache.kafka.server.log.remote.metadata.storage.ConsumerManager.waitTillConsumptionCatchesUp(ConsumerManager.java:80)
org.apache.kafka.server.log.remote.metadata.storage.TopicBasedRemoteLogMetadataManager.lambda$doPublishMetadata$0(TopicBasedRemoteLogMetadataManager.java:165)
java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

junrao replied:
Thanks for the explanation. Semantically, for the CompletableFuture returned from doPublishMetadata(), when do we expect it to complete? The implementation completes it after the metadata is acked in the producer. However, I thought it should be completed after the consumer has caught up on the offset? It would be useful to document this in the API clearly.

satishd replied:
`produceFuture` is completed after `ProducerRecordMetadata` is completed. But `doPublishMetadata` takes `produceFuture` and composes with `.thenApplyAsync()` and returns the `CompletableFuture` which will be completed only after the `consumerManager.waitTillConsumptionCatchesUp(recordMetadata);` is returned. 
So, the returned `CompletableFuture` from `doPublishMetadata` is completed only after the consumer is caughtup until the produced record offset. I will document it in the code.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments below.

satishd committed:
Minor cleanup and addressed review comments.

ccding commented on storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java:
```diff
@@ -62,16 +63,17 @@
      * <pre>
      * +---------------------+            +----------------------+
-     * |COPY_SEGMENT_STARTED |-----------&gt;|COPY_SEGMENT_FINISHED |
+     * |COPY_SEGMENT_STARTED |----------->|COPY_SEGMENT_FINISHED |
```
Can you verify if this could result in a correct HTML doc?

satishd replied:
I verified that the correct HTML is generated in the javadoc.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -50,37 +51,38 @@ public ProducerManager(TopicBasedRemoteLogMetadataManagerConfig rlmmConfig,
+                                               serde.serialize(remoteLogMetadata)), callback);
+        } catch (Exception ex) {
+            future.completeExceptionally(ex);
         }
```
do we want to remove printing the topic id in the exception?

satishd replied:
Throwing exception is removed as you can see in the diff.

satishd commented:
Thanks @junrao for the review comments, addressed them with the latest commit.

satishd committed:
Updated with a few more comments.

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM
